### PR TITLE
Move code for local git repository out of Archiver 

### PR DESCRIPTION
**Pull request by smola**
* Add TemporaryCloner and TemporaryRepository abstractions
  to clone a remote repository to a temporary local storage
  and use it to push to different URLs.

* Fetch all references with `refs/` prefix.

### CODE REVIEW

**Comment by smola** [comment]
Merge https://github.com/src-d/borges/pull/98 first.

File: archiver.go [hunk 1]
```diff
@@ -270,7 +202,7 @@ func (a *Archiver) pushChangesToRootedRepositories(j *Job, r *model.Repository,
 	return checkFailedInits(changes, failedInits)
 }
 
-func (a *Archiver) pushChangesToRootedRepository(r *model.Repository, lr *git.Repository, ic model.SHA1, changes []*Command) error {
+func (a *Archiver) pushChangesToRootedRepository(r *model.Repository, lr TemporaryRepository, ic model.SHA1, changes []*Command) error {
```
Comment (ajnavarro) [inline archiver.go]: Could you change lr for tr?

File: archiver_test.go [hunk 2]
```diff
@@ -55,6 +54,10 @@ func (s *ArchiverSuite) TestReferenceUpdate() {
 	}
 }
 
+func (s *ArchiverSuite) TestDoWithFailingClones() {
+	s.T().Skip("foo")
```
Comment (ajnavarro) [inline archiver_test.go]: Can you check this?

File: fixtures_test.go [hunk 3]
```diff
@@ -232,6 +232,14 @@ func (f *ChangesFixture) setReferences(r *git.Repository, refs []*model.Referenc
 		}
 	}
 
+	// FIXME: We have some problems in go-git cloning repositories that do not
+	// have HEAD. We add one to avoid these problems.
+	_, err := r.Storer.Reference(plumbing.HEAD)
```
Comment (ajnavarro) [inline fixtures_test.go]: Actually, at the end we are fetching, right? This is not needed.

File: git.go [hunk 4]
```diff
@@ -32,6 +57,10 @@ func (r gitReferencer) References() ([]*model.Reference, error) {
 			return nil
 		}
 
+		if ref.Name() == plumbing.HEAD {
```
Comment (ajnavarro) [inline git.go]: This is not needed either.

File: git.go [hunk 5]
```diff
@@ -75,3 +104,124 @@ func rootCommits(r *git.Repository, from plumbing.Hash) ([]model.SHA1, error) {
...
+		Name: "origin",
+		URL:  endpoint,
+	})
+	if err != nil {
+		_ = util.RemoveAll(b.TempFilesystem, dir)
+		return nil, err
+	}
+
+	o := &git.FetchOptions{
+		RefSpecs: []config.RefSpec{config.RefSpec("refs/*:refs/*")},
```
Comment (ajnavarro) [inline git.go]: Can you put the refspec in a constant?

**Comment by ajnavarro** [comment]
Also, can you change the PR description? It says that we will use Clone instead of Fetch, but is not true.

**Pull request merged**